FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
+
+/* - - - - - - sampleHousekeepingData - - - - - - *
+ * Usage:
+ *  samples temperature and power data and stores it in hkData,
+ *  checks for out-of-bounds values and logs faults if found
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void sampleHousekeepingData()
+{
+    // read thermistor voltage values
+    float opticsTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_OPTICS_THERM);
+    float analogTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_ANALOG_THERM);
+    float digitalTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_DIGITAL_THERM);
+    
+    // update housekeeping data
+    latestHkSample.opticsTemp = voltageToTemp(opticsTempVoltage); 
+    latestHkSample.analogTemp = voltageToTemp(analogTempVoltage);
+    latestHkSample.digitalTemp = voltageToTemp(digitalTempVoltage);
+    latestHkSample.analogCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_AREG_CURR);
+    latestHkSample.digitalCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_CURR);   
+    latestHkSample.digitalRegPG = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_PG);
+    hkData[hkIndex] = latestHkSample; // update data array
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
+    else if (latestHkSample.opticsTemp < OPTICS_TEMP_MIN_SAFE) { } // optics temp too low
+
+    if (latestHkSample.analogTemp > BOARD_TEMP_MAX_SAFE) { } // analog board temp too high
+    else if (latestHkSample.analogTemp < BOARD_TEMP_MIN_SAFE) { } // analog board temp too low
+
+    if (latestHkSample.digitalTemp > BOARD_TEMP_MAX_SAFE) { } // digital board temp too high
+    else if (latestHkSample.digitalTemp < BOARD_TEMP_MIN_SAFE) { } // digital board temp too low
+
+    if (latestHkSample.digitalRegPG > PG_VOLTAGE_MAX_EXPECTED) { } // digital regulator PG signal too high
+    else if (latestHkSample.digitalRegPG < PG_VOLTAGE_MIN_EXPECTED) { } // digital regulator PG signal too low
+
+}
+
+/* - - - - - - voltageToTemp - - - - - - *
+ * Usage:
+ *  Interpolates and returns the corresponding temperature for a given thermistor voltage reading
+ *  based on the data in thermLookup. 
+ *  Will extrapolate beyond max and min entries in thermLookup if needed
+ * 
+ * Inputs:
+ *  voltage - thermistor voltage
+ *  
+ * Outputs:
+ *  temperature - celsius, the temperature corresponding to thermistor voltage reading
+ */
+float voltageToTemp(float voltage)
+{
+    int thermLookupSize = sizeof(thermLookup) / sizeof(thermLookup[0]); // size of lookup table
+    
+    // find lower adjacent entry
+    int i = 0;
+    while (thermLookup[i].voltage > voltage && i < thermLookupSize - 1) { i++; }
```

Comment:
This looks good. If these calculations do prove to be too intense at the required rate, you can just compare raw voltage values (just take the corresponding voltages at the triggering temperatures). Then we provide good documentation on how to convert voltages to temperatures on the ground for science.